sdk/rust/oak_utils/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use std::collections::HashMap;
+use std::fs;
+use std::io;
+use std::path::Path;
+
+#[cfg(test)]
+mod tests;
+
+// This function uses `protoc_rust` to generate Rust `proto` files in a temporary directory,
+// checks previously generated files and updates them if their contents have changed.
+// This is a workaround for `protoc_rust` that always updates files, thus provoking recompilation
+// of all dependent targets.
+// Function doesn't support nested directories since `protoc_rust` doesn't generate them.
+pub fn run_protoc_rust(args: protoc_rust::Args) -> io::Result<()> {
+    let out_path = Path::new(args.out_dir);
+
+    // Create a temporary directory.
+    let temp_dir = tempfile::tempdir()?;
+    let temp_path = temp_dir.path();
+
+    // Generate Rust `proto` files in the temporary directory.
+    let mut temp_args = args;
+    temp_args.out_dir = temp_path.to_str().expect("Temporary path error");
+    protoc_rust::run(temp_args)?;
+
+    // Copy updated Rust `proto` files to the `out_path`.
+    copy_updated_files(temp_path, out_path)?;
+
+    Ok(())
+}
+
+// This function is identical to `run_protoc_rust` but uses `protoc_rust_grpc` instead.
+pub fn run_protoc_rust_grpc(args: protoc_rust_grpc::Args) -> io::Result<()> {
+    let out_path = Path::new(args.out_dir);
+
+    // Create a temporary directory.
+    let temp_dir = tempfile::tempdir()?;
+    let temp_path = temp_dir.path();
+
+    // Generate Rust `grpc` files in the temporary directory.
+    let mut temp_args = args;
+    temp_args.out_dir = temp_path.to_str().expect("Temporary path error");
+    protoc_rust_grpc::run(temp_args)?;
+
+    // Copy updated Rust `grpc` files to the `out_path`.
+    copy_updated_files(temp_path, out_path)?;
+
+    Ok(())
+}
+
+// This function traverses `dir` and gets a `HashMap` of filenames and their contents.
```

Comment:
Perhaps remove all the "This function" parts? I think this would read fine as "Traverses `dir` and produces a `HashMap` [...]".

See examples here: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts to see how std library documentation is usually written.

Reply:
Done